bpabi32/bpabi32.rst

```diff
@@ -0,0 +1,2955 @@
+straightforwardly because the dynamic linker looks up symbol names while
+relocating an executable file. Pre-emption follows naturally from the way
+the symbol table for a process is constructed.
+
+In general, a DLL-based system must perform the required symbol table
+juggling at static link time when the application that will generate a
+process is created. This is the only point at which all the DLLs to be
+linked into the process are known and symbol tables are still available. By
+dynamic link time, otherwise global symbol names will have been translated
+into DLL-local ordinal numbers, and no process-wide symbol table will be
+built by the loader.
+
+So, dynamic vague linkage must be resolved when an application is created,
+and recorded in the resulting BPABI executable file in a form that
+post-linkers can translate into platform-specific dynamic relocation
+directives. This observation motivates the specification given in
+`Encoding symbol pre-emption in BPABI executable files`_.
+
+Relating executables and shared objects to executable files
+-----------------------------------------------------------
+
+Terminology
+^^^^^^^^^^^
+
+[ELF_] already gives precise definitions to the words segment and section.
+Section is not relevant to the executable view of an ELF file, so it is not
+available for more general usage here. We use (program) segment with its
+strict ELF meaning to denote a contiguous part of an executable file read by
+a loader or dynamic linker. Where segment explicitly or implicitly has type
+PT_LOAD, it refers a contiguous part of an executable file loaded into
+memory to initialize an executable image.
+
+We also need a word to describe part of an executable image loaded into
+virtual memory. We use the word region, familiar to users of Arm development
+tools. An (execution) region contains the minimum needed to support
+execution of the program on a bare platform.  A (program) segment usually
+contains one (execution) region. On some platforms it also contains other
+platform-specific data associated with loading or dynamic linking.
+
+Finally, we need a term for part of an executable file that serves a
+distinct, coherent purpose. We use the word component. A segment is also a
+component. So is the content of a region when it is stored in an executable
+file. On some platforms there are components that are not part of any region
+or segment. Clearly, a component can have sub-components.
+
+Logical content of an executable file
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The figure below depicts a typical set of components present in an
+executable file. Not all components are present in all platform-specific
+formats, and the encoding of dynamic linking data varies greatly between
+platforms. (The dynamic components shown in the figure are a mélange of DLL
+and SVr4 components).
+
+.. rubric:: Components of a typical executable file and the relationships between them
+
+.. figure:: bpabi32-exe-file.svg
```

Comment:
In the svg, the "Dynamic components (contain dynamic linking data)" text is placed halfway out of the brown box, also making it a bit harder to read.

Reply:
Good catch. I got into a muddle with Inkscape and had to reset a bunch of font settings at the last minute, and the UI for doing that also resets the text alignments unwantedly and I had to put them all back by hand. Missed one.